frida-gum/src/module_map.rs

```diff
@@ -0,0 +1,179 @@
+                        .eq(&context.name))
+            {
+                context.details = gum_sys::gum_module_details_copy(details);
+                return 0;
+            }
+
+            1
+        }
+
+        let mut context = SaveModuleDetailsByNameContext {
+            name,
+            details: std::ptr::null_mut(),
+        };
+
+        unsafe {
+            gum_sys::gum_process_enumerate_modules(
+                Some(save_module_details_by_name),
+                &mut context as *mut _ as *mut c_void,
+            );
+        }
+
+        if context.details.is_null() {
+            None
+        } else {
+            Some(Self::from_raw(context.details))
+        }
+    }
+
+    /// Get the name of this module
+    pub fn name(&self) -> String {
+        unsafe {
+            CStr::from_ptr((*self.module_details).name)
+                .to_string_lossy()
+                .to_string()
+        }
+    }
+
+    /// Get the path of this module
+    pub fn path(&self) -> String {
+        unsafe {
+            CStr::from_ptr((*self.module_details).path)
+                .to_string_lossy()
+                .to_string()
+        }
+    }
+
+    /// Get the range of this module
+    pub fn range(&self) -> MemoryRange {
+        MemoryRange::from_raw(unsafe { (*self.module_details).range })
+    }
+}
+
+pub struct ModuleMap {
+    pub(crate) module_map: *mut gum_sys::GumModuleMap,
+}
+
+impl ModuleMap {
+    pub(crate) fn from_raw(module_map: *mut gum_sys::GumModuleMap) -> Self {
+        Self { module_map }
+    }
+
+    /// Create a new [`ModuleMap`]
+    pub fn new() -> Self {
+        Self::from_raw(unsafe { gum_sys::gum_module_map_new() })
+    }
+
+    /// Create a new [`ModuleMap`] with a filter function
+    pub fn new_with_filter(filter: &mut dyn FnMut(ModuleDetails) -> bool) -> Self {
+        unsafe extern "C" fn module_map_filter(
+            details: *const gum_sys::_GumModuleDetails,
+            callback: *mut c_void,
+        ) -> i32 {
+            let callback = &mut *(callback as *mut Box<&mut dyn FnMut(ModuleDetails) -> bool>);
+            if (callback)(ModuleDetails::from_raw(details)) {
+                1
+            } else {
+                0
+            }
+        }
+        Self::from_raw(unsafe {
+            gum_sys::gum_module_map_new_filtered(
+                Some(module_map_filter),
+                &mut Box::new(filter) as *mut _ as *mut c_void,
+                None,
+            )
+        })
+    }
+
+    /// Create a new [`ModuleMap`] from a list of names
+    pub fn new_from_names(names: &[&str]) -> Self {
+        Self::new_with_filter(&mut |details: ModuleDetails| {
+            for name in names {
+                if (name.starts_with('/') && details.path().eq(name))
+                    || (name.contains('/')
+                        && details.name().eq(Path::new(name)
+                            .file_name()
+                            .unwrap()
```

Comment:
I think `unwrap`ing internally is not the best form. This ties in with the previous custom Error type I mentioned. Will fix this up later.

Reply:
Ok. I'll leave it for now.